File: ts/src/OnePieceSDK.ts
// OnePiece Ts SDK

import { BoatEntity } from './entity/BoatEntity'
import { BowEntity } from './entity/BowEntity'
import { ChapterEntity } from './entity/ChapterEntity'
import { CharacterEntity } from './entity/CharacterEntity'
import { CrewEntity } from './entity/CrewEntity'
import { DialEntity } from './entity/DialEntity'
import { EpisodeEntity } from './entity/EpisodeEntity'
import { FilmEntity } from './entity/FilmEntity'
import { FruitEntity } from './entity/FruitEntity'
import { GearEntity } from './entity/GearEntity'
import { HakiEntity } from './entity/HakiEntity'
import { LocationEntity } from './entity/LocationEntity'
import { SagaEntity } from './entity/SagaEntity'
import { SwordEntity } from './entity/SwordEntity'
import { TechniqueEntity } from './entity/TechniqueEntity'
import { VolumeEntity } from './entity/VolumeEntity'

export type * from './OnePieceTypes'


import { inspect } from 'node:util'

import type { Context, Feature } from './types'

import { config } from './Config'
import { OnePieceEntityBase } from './OnePieceEntityBase'
import { Utility } from './utility/Utility'


import { BaseFeature } from './feature/base/BaseFeature'


const stdutil = new Utility()


class OnePieceSDK {
  _mode: string = 'live'
  _options: any
  _utility = new Utility()
  _features: Feature[]
  _rootctx: Context

  constructor(options?: any) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath
    const items = struct.items

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    items(this._options.feature, (fitem: [string, any]) => {
      const fname = fitem[0]
      const fopts = fitem[1]
      if (fopts.active) {
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    })

    if (null != this._options.extend) {
      for (let f of this._options.extend) {
        featureAdd(this._rootctx, f)
      }
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs?: any) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx: Context = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec: any = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  async direct(fetchargs?: any) {
    const utility = this._utility
    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx: Context = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json: any = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err: any) {
      return { ok: false, err }
    }
  }



  _boat?: BoatEntity

  // Idiomatic facade: `client.boat.list()` / `client.boat.load({ id })`.
  get boat(): BoatEntity {
    return (this._boat ??= new BoatEntity(this, undefined))
  }

  /** @deprecated Use `client.boat` instead. */
  Boat(data?: any) {
    const self = this
    return new BoatEntity(self,data)
  }


  _bow?: BowEntity

  // Idiomatic facade: `client.bow.list()` / `client.bow.load({ id })`.
  get bow(): BowEntity {
    return (this._bow ??= new BowEntity(this, undefined))
  }

  /** @deprecated Use `client.bow` instead. */
  Bow(data?: any) {
    const self = this
    return new BowEntity(self,data)
  }


  _chapter?: ChapterEntity

  // Idiomatic facade: `client.chapter.list()` / `client.chapter.load({ id })`.
  get chapter(): ChapterEntity {
    return (this._chapter ??= new ChapterEntity(this, undefined))
  }

  /** @deprecated Use `client.chapter` instead. */
  Chapter(data?: any) {
    const self = this
    return new ChapterEntity(self,data)
  }


  _character?: CharacterEntity

  // Idiomatic facade: `client.character.list()` / `client.character.load({ id })`.
  get character(): CharacterEntity {
    return (this._character ??= new CharacterEntity(this, undefined))
  }

  /** @deprecated Use `client.character` instead. */
  Character(data?: any) {
    const self = this
    return new CharacterEntity(self,data)
  }


  _crew?: CrewEntity

  // Idiomatic facade: `client.crew.list()` / `client.crew.load({ id })`.
  get crew(): CrewEntity {
    return (this._crew ??= new CrewEntity(this, undefined))
  }

  /** @deprecated Use `client.crew` instead. */
  Crew(data?: any) {
    const self = this
    return new CrewEntity(self,data)
  }


  _dial?: DialEntity

  // Idiomatic facade: `client.dial.list()` / `client.dial.load({ id })`.
  get dial(): DialEntity {
    return (this._dial ??= new DialEntity(this, undefined))
  }

  /** @deprecated Use `client.dial` instead. */
  Dial(data?: any) {
    const self = this
    return new DialEntity(self,data)
  }


  _episode?: EpisodeEntity

  // Idiomatic facade: `client.episode.list()` / `client.episode.load({ id })`.
  get episode(): EpisodeEntity {
    return (this._episode ??= new EpisodeEntity(this, undefined))
  }

  /** @deprecated Use `client.episode` instead. */
  Episode(data?: any) {
    const self = this
    return new EpisodeEntity(self,data)
  }


  _film?: FilmEntity

  // Idiomatic facade: `client.film.list()` / `client.film.load({ id })`.
  get film(): FilmEntity {
    return (this._film ??= new FilmEntity(this, undefined))
  }

  /** @deprecated Use `client.film` instead. */
  Film(data?: any) {
    const self = this
    return new FilmEntity(self,data)
  }


  _fruit?: FruitEntity

  // Idiomatic facade: `client.fruit.list()` / `client.fruit.load({ id })`.
  get fruit(): FruitEntity {
    return (this._fruit ??= new FruitEntity(this, undefined))
  }

  /** @deprecated Use `client.fruit` instead. */
  Fruit(data?: any) {
    const self = this
    return new FruitEntity(self,data)
  }


  _gear?: GearEntity

  // Idiomatic facade: `client.gear.list()` / `client.gear.load({ id })`.
  get gear(): GearEntity {
    return (this._gear ??= new GearEntity(this, undefined))
  }

  /** @deprecated Use `client.gear` instead. */
  Gear(data?: any) {
    const self = this
    return new GearEntity(self,data)
  }


  _haki?: HakiEntity

  // Idiomatic facade: `client.haki.list()` / `client.haki.load({ id })`.
  get haki(): HakiEntity {
    return (this._haki ??= new HakiEntity(this, undefined))
  }

  /** @deprecated Use `client.haki` instead. */
  Haki(data?: any) {
    const self = this
    return new HakiEntity(self,data)
  }


  _location?: LocationEntity

  // Idiomatic facade: `client.location.list()` / `client.location.load({ id })`.
  get location(): LocationEntity {
    return (this._location ??= new LocationEntity(this, undefined))
  }

  /** @deprecated Use `client.location` instead. */
  Location(data?: any) {
    const self = this
    return new LocationEntity(self,data)
  }


  _saga?: SagaEntity

  // Idiomatic facade: `client.saga.list()` / `client.saga.load({ id })`.
  get saga(): SagaEntity {
    return (this._saga ??= new SagaEntity(this, undefined))
  }

  /** @deprecated Use `client.saga` instead. */
  Saga(data?: any) {
    const self = this
    return new SagaEntity(self,data)
  }


  _sword?: SwordEntity

  // Idiomatic facade: `client.sword.list()` / `client.sword.load({ id })`.
  get sword(): SwordEntity {
    return (this._sword ??= new SwordEntity(this, undefined))
  }

  /** @deprecated Use `client.sword` instead. */
  Sword(data?: any) {
    const self = this
    return new SwordEntity(self,data)
  }


  _technique?: TechniqueEntity

  // Idiomatic facade: `client.technique.list()` / `client.technique.load({ id })`.
  get technique(): TechniqueEntity {
    return (this._technique ??= new TechniqueEntity(this, undefined))
  }

  /** @deprecated Use `client.technique` instead. */
  Technique(data?: any) {
    const self = this
    return new TechniqueEntity(self,data)
  }


  _volume?: VolumeEntity

  // Idiomatic facade: `client.volume.list()` / `client.volume.load({ id })`.
  get volume(): VolumeEntity {
    return (this._volume ??= new VolumeEntity(this, undefined))
  }

  /** @deprecated Use `client.volume` instead. */
  Volume(data?: any) {
    const self = this
    return new VolumeEntity(self,data)
  }




  static test(testoptsarg?: any, sdkoptsarg?: any) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new OnePieceSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts?: any, sdkopts?: any) {
    return OnePieceSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'OnePiece' }
  }

  toString() {
    return 'OnePiece ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = OnePieceSDK


export {
  stdutil,

  BaseFeature,
  OnePieceEntityBase,

  OnePieceSDK,
  SDK,
}

File: py/onepiece_sdk.py
# OnePiece SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import OnePieceUtility
from core.spec import OnePieceSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import OnePieceBaseFeature
from features import _make_feature


class OnePieceSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = OnePieceUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return OnePieceUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = OnePieceSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def Boat(self, data=None) -> "BoatEntity":
        """Entity factory: client.Boat().list() / client.Boat().load({"id": ...})."""
        from entity.boat_entity import BoatEntity
        return BoatEntity(self, data)


    def Bow(self, data=None) -> "BowEntity":
        """Entity factory: client.Bow().list() / client.Bow().load({"id": ...})."""
        from entity.bow_entity import BowEntity
        return BowEntity(self, data)


    def Chapter(self, data=None) -> "ChapterEntity":
        """Entity factory: client.Chapter().list() / client.Chapter().load({"id": ...})."""
        from entity.chapter_entity import ChapterEntity
        return ChapterEntity(self, data)


    def Character(self, data=None) -> "CharacterEntity":
        """Entity factory: client.Character().list() / client.Character().load({"id": ...})."""
        from entity.character_entity import CharacterEntity
        return CharacterEntity(self, data)


    def Crew(self, data=None) -> "CrewEntity":
        """Entity factory: client.Crew().list() / client.Crew().load({"id": ...})."""
        from entity.crew_entity import CrewEntity
        return CrewEntity(self, data)


    def Dial(self, data=None) -> "DialEntity":
        """Entity factory: client.Dial().list() / client.Dial().load({"id": ...})."""
        from entity.dial_entity import DialEntity
        return DialEntity(self, data)


    def Episode(self, data=None) -> "EpisodeEntity":
        """Entity factory: client.Episode().list() / client.Episode().load({"id": ...})."""
        from entity.episode_entity import EpisodeEntity
        return EpisodeEntity(self, data)


    def Film(self, data=None) -> "FilmEntity":
        """Entity factory: client.Film().list() / client.Film().load({"id": ...})."""
        from entity.film_entity import FilmEntity
        return FilmEntity(self, data)


    def Fruit(self, data=None) -> "FruitEntity":
        """Entity factory: client.Fruit().list() / client.Fruit().load({"id": ...})."""
        from entity.fruit_entity import FruitEntity
        return FruitEntity(self, data)


    def Gear(self, data=None) -> "GearEntity":
        """Entity factory: client.Gear().list() / client.Gear().load({"id": ...})."""
        from entity.gear_entity import GearEntity
        return GearEntity(self, data)


    def Haki(self, data=None) -> "HakiEntity":
        """Entity factory: client.Haki().list() / client.Haki().load({"id": ...})."""
        from entity.haki_entity import HakiEntity
        return HakiEntity(self, data)


    def Location(self, data=None) -> "LocationEntity":
        """Entity factory: client.Location().list() / client.Location().load({"id": ...})."""
        from entity.location_entity import LocationEntity
        return LocationEntity(self, data)


    def Saga(self, data=None) -> "SagaEntity":
        """Entity factory: client.Saga().list() / client.Saga().load({"id": ...})."""
        from entity.saga_entity import SagaEntity
        return SagaEntity(self, data)


    def Sword(self, data=None) -> "SwordEntity":
        """Entity factory: client.Sword().list() / client.Sword().load({"id": ...})."""
        from entity.sword_entity import SwordEntity
        return SwordEntity(self, data)


    def Technique(self, data=None) -> "TechniqueEntity":
        """Entity factory: client.Technique().list() / client.Technique().load({"id": ...})."""
        from entity.technique_entity import TechniqueEntity
        return TechniqueEntity(self, data)


    def Volume(self, data=None) -> "VolumeEntity":
        """Entity factory: client.Volume().list() / client.Volume().load({"id": ...})."""
        from entity.volume_entity import VolumeEntity
        return VolumeEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "OnePieceSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.boat_entity import BoatEntity
    from entity.bow_entity import BowEntity
    from entity.chapter_entity import ChapterEntity
    from entity.character_entity import CharacterEntity
    from entity.crew_entity import CrewEntity
    from entity.dial_entity import DialEntity
    from entity.episode_entity import EpisodeEntity
    from entity.film_entity import FilmEntity
    from entity.fruit_entity import FruitEntity
    from entity.gear_entity import GearEntity
    from entity.haki_entity import HakiEntity
    from entity.location_entity import LocationEntity
    from entity.saga_entity import SagaEntity
    from entity.sword_entity import SwordEntity
    from entity.technique_entity import TechniqueEntity
    from entity.volume_entity import VolumeEntity
